### ISSUE TITLE
Implement SensorKit on iOS

### ISSUE THREAD

**Issue by avaidyam** [BIDMCDigitalPsychiatry/LAMP-platform#139]
Implement the new `SensorKit` framework on iOS 14 -- on hold until we receive approval from Apple (see "Required Entitlements" below).

### `SRSensor`s
 - `accelerometer`: A sensor that provides acceleration motion data.
 - `rotationRate`: A sensor that provides rotation motion data.
 - `pedometerData`: A sensor that provides information about the user's steps.
 - `ambientLightSensor`: A sensor that provides ambient light information.
     - `chromaticity{x, y}`: A coordinate pair that describes the sample’s light brightness and tint.
     - `lux`: An object that describes the sample’s luminous flux.
     - `placement`: The light’s location relative to the sensor.
 - `onWristState`: A sensor that describes the watch’s position on the wrist.
     - `crownOrientation`: A value that indicates the direction the Digital Crown faces with respect to the wearer.
     - `onWrist`: A value that indicates whether the watch is on the user’s wrist.
     - `wristLocation`: A value that indicates the wrist where the user wears the watch.
 - `visits`: A sensor that provides information about the user’s frequently visited locations.
     - `arrivalDateInterval`: A range of time within which the user arrives at a location of interest.
     - `departureDateInterval`: A range of time within which the user departs from a location of interest.
     - `distanceFromHome`: The location’s distance from the home-category location.
     - `locationCategory`: `gym`, `home`, `school`, `work`, `unknown`.
 - `keyboardMetrics`: A sensor that provides information about the user’s keyboard usage.
     - `duration`: The duration that the report spans.
     - `keyboardIdentifier`: The identifier of the keyboard in the keyboard list.
     - `version`: The version of keyboard metrics.
     - `width`: The width, in millimeters, of the keyboard in the report.
     - `height`: The height, in millimeters, of the keyboard in the report.
     - `totalWords`: The total number of typed words for the keyboard.
     - `totalAlteredWords`: The total number of altered words for the keyboard.
     - `totalTaps`: The total number of taps for the keyboard.
     - `totalDrags`: The total number of drags for the keyboard.
     - `totalDeletes`: The total number of deletions for the keyboard.
     - `totalEmojis`: The total number of emojis for the keyboard.
     - `totalPaths`: The total number of completed paths for the keyboard.
     - `totalPathTime`: The total time to complete paths for the keyboard.
     - `totalPathLength`: The total length of completed paths for the keyboard.
     - `totalAutoCorrections`: The total number of autocorrections for the keyboard.
     - `totalSpaceCorrections`: The total number of space corrections for the keyboard.
     - `totalRetroCorrections`: The total number of retro corrections for the keyboard.
     - `totalTranspositionCorrections`: The total number of transposition corrections for the keyboard.
     - `totalInsertKeyCorrections`: The total number of insert key corrections for the keyboard.
     - `totalSkipTouchCorrections`: The total number of skip touch corrections for the keyboard.
     - `totalNearKeyCorrections`: The total number of near key corrections for the keyboard.
     - `totalSubstitutionCorrections`: The total number of substitution corrections for the keyboard.
     - `totalHitTestCorrections`: The total number of hit test corrections for the keyboard.
     - `totalTypingDuration`: The total amount of time typing for the keyboard.
     - `touchDownUp`: The duration between touch down to touch up for any key.
     - `spaceTouchDownUp`: The duration between touch down and touch up of all Space bar events for the keyboard.
     - `deleteTouchDownUp`: The duration between touch down and touch up of all Delete key events for the keyboard.
     - `shortWordCharKeyTouchDownUp`: The duration between touch down and touch up of all character keys in short words for the keyboard.
     - `touchDownDown`: The duration between touch down to touch down for any key.
     - `charKeyToPrediction`: The duration between touch up on a character key and touch down on a word in the prediction bar.
     - `shortWordCharKeyToCharKey`: The duration between touch up on a character key and touch down on any sequential character key in a short word.
     - `charKeyToAnyTapKey`: The duration between touch up on a character key and touch down on the next sequential key.
     - `anyTapToCharKey`: The duration between touch up of any key and touch down on a sequential character key.
     - `spaceToCharKey`: The duration between touch up of the Space bar and touch down of a sequential character key.
     - `charKeyToSpaceKey`: The duration between touch up of a character key and touch down of a sequential Space bar.
     - `spaceToDeleteKey`: The duration between touch up of the Space bar and touch down of a sequential Delete key.
     - `deleteToSpaceKey`: The duration between touch up of the Delete key and touch down of a sequential Space bar.
     - `spaceToSpaceKey`: The duration between touch up of the Space bar and touch down of a sequential Space bar.
     - `spaceToShiftKey`: The duration between touch up of the Space bar and touch down of a sequential Shift key.
     - `spaceToPlaneChangeKey`: The duration between touch up of the Space bar and touch down of a sequential plane change key.
     - `spaceToPredictionKey`: The duration between touch up of the Space bar and touch down of a sequential selection from the prediction bar.
     - `deleteToCharKey`: The duration between touch up of the Delete key and touch down of a sequential character key.
     - `charKeyToDelete`: The duration between touch up of a character key and touch down of a sequential Delete key.
     - `deleteToDelete`: The duration between touch up of the Delete key and touch down of a sequential Delete key.
     - `deleteToDeletes`: The duration between touch up of the Delete key and touch down of a sequential Delete key, for an entire word.
     - `deleteToShiftKey`: The duration between touch up of the Delete key and touch down of a sequential Shift key.
     - `deleteToPlaneChangeKey`: The duration between touch up of the Delete key and touch down of a sequential plane change key.
     - `anyTapToPlaneChangeKey`: The duration between touch up of any key and touch down on a plane change key.
     - `planeChangeToAnyTap`: The duration between touch up on a plane change key and touch down on the next sequential key.
     - `charKeyToPlaneChangeKey`: The duration between touch up of a character key and touch down of a sequential plane change key.
     - `planeChangeKeyToCharKey`: The duration between touch up of a plane change key and touch down of any key.
     - `deleteToPath`: The duration between touch up of the Delete key and touch down of a sequential path.
     - `pathToDelete`: The duration between touch up of the Delete key and touch down of a continuous path.
     - `spaceToPath`: The duration between touch up of the Space bar and touch down to begin a sequential path.
     - `pathToSpace`: The duration between touch up of a path and touch down of a sequential Space bar.
     - `pathToPath`: The duration between touch up of a path and touch down of a sequential path.
     - `longWordTouchDownUp`: The duration between touch down and touch up of the character keys of all the long words for the keyboard.
     - `longWordTouchDownDown`: The duration between touch down and touch down of the character keys of all the long words for the keyboard.
     - `longWordUpErrorDistance`: The distance from the touch up to the center of the intended key of the characters of a long word.
     - `longWordDownErrorDistance`: The distance from the touch down to the center of the intended key of the characters of a long word.
     - `upErrorDistance`: The distance from the touch up to the center of any key.
     - `downErrorDistance`: The distance from the touch down to the center of any key.
     - `spaceUpErrorDistance`: The distance from the touch up to the right centroid of the Space bar.
     - `spaceDownErrorDistance`: The distance from the touch down to the right centroid of the Space bar.
     - `deleteUpErrorDistance`: The distance from the touch up to the center of the Delete key.
     - `deleteDownErrorDistance`: The distance from the touch down to the center of the Delete key.
     - `shortWordCharKeyUpErrorDistance`: The distance from the touch up to the center of the intended key of a character in a short word.
     - `shortWordCharKeyDownErrorDistance`: The distance from the touch down to the center of the intended key of a character in a short word.
     - `pathErrorDistanceRatio`: Sample values of the ratio of error distance between the intended and actual path.
 - `deviceUsageReport`: A sensor that provides information about the user’s device usage.
     - `duration`: The duration that the report spans.
     - `totalScreenWakes`: The total number of screen wakes for the device.
     - `totalUnlocks`: The total number of unlocks for the device.
     - `totalUnlockDuration`: The duration of time the device is in an unlocked state.
     - `applicationUsageByCategory`: The usage time of apps per category.
     - `notificationUsageByCategory`: The frequency of notifications per category.
     - `webUsageByCategory`: The amount of time the user accesses domains per category.
 - `messagesUsageReport`: A sensor that provides information about the user’s use of the Messages app.
     - `duration`: The duration that the report spans.
     - `totalIncomingMessages`: The number of messages the user receives.
     - `totalOutgoingMessages`: The number of messages the user sends.
     - `totalUniqueContacts`: The user’s number of contacts.
 - `phoneUsageReport`: A sensor that reports the amount of time that the user is on phone calls.
     - `duration`: The duration that the report spans.
     - `totalIncomingCalls`: The number of calls the user receives.
     - `totalOutgoingCalls`: The number of calls the user makes.
     - `totalPhoneCallDuration`: The total duration of all calls.
     - `totalUniqueContacts`: The user’s number of contacts.

### Implementation Steps
1. [Configure `LAMP-core-ios`.](https://developer.apple.com/documentation/sensorkit/configuring_your_project_for_sensor_reading)
2. [Set up `SRSensorReader`.](https://developer.apple.com/documentation/sensorkit/srsensorreader)
3. [Start an `SRFetchRequest`.](https://developer.apple.com/documentation/sensorkit/srfetchrequest)

### Required Entitlements 
- [ ]  `com.apple.developer.sensorkit.reader.allow`
    - [ ]  `motion-accelerometer` → higher frequency (?) accelerometer
    - [ ]  `motion-rotation-rate` → higher frequency (?) gyroscope
    - [ ]  `pedometer` → higher frequency (?) pedometer
    - [ ]  `on-wrist` → apple watch and airpods orientation/motion sensors
    - [ ]  `ambient-light-sensor` → ambient lighting conditions
    - [ ]  `messages-usage` → SMS sent/received
    - [ ]  `phone-usage` → phone call sent/received
    - [ ]  `visits` → significant locations & "home time"
    - [ ]  `device-usage` → screen time, notifications, pickups (per app)
    - [ ]  `keyboard-metrics` → keyboard usage metrics

**Comment by ZCOEngineer** [comment]
hi @SuruPa00 
Jijo will start on this soon, We shall update you on the ETA

**Comment by ZCOEngineer** [comment]
@SuruPa00 we are working in this and expected release date is by Dec-08

**Comment by ZCOEngineer** [comment]
Hi @SuruPa00 
We are ready for developer testing SensorKit integration. We need to add the related entitlement to the provisioning profile. But the option for adding this is still not available in the LAMP developer account. Could you check if there is anything regarding this to be done on your part to get it available?

**Comment by jijopulikkottil** [comment]
@avaidyam  same question [here](https://developer.apple.com/forums/thread/655277) . 
FYI we are using the team 'John Torous' under the apple account 'John Torous'

**Comment by avaidyam** [comment]
I'll take a look and let you know shortly if there's any other steps to be taken on your end.

**Comment by ZCOEngineer** [comment]
@avaidyam Any update on this one Aditya, We are still waiting on doing the testing.

**Comment by jijopulikkottil** [comment]
HI @avaidyam, The new entitlements are working fine. It is asking permissions. Hope we can start its integration.

**Comment by avaidyam** [comment]
@jijopulikkottil Yes, please begin testing and integration for staging. For now, please restrict these sensor types to only the server addresses from the domain group `*.lamp.digital`.

**Comment by jijopulikkottil** [comment]
Ok, we just tested few sensors with a sample app and able to fetch data after 2 days.  The `rotation rate` sensor doesn't  return any data even after 3 days. But as per above [comment](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/139#issuecomment-944364602), we are not going to fetch `Core-Motion` sensors. So not bothering about that now.

Hope we should add this implementation to LAMP-Swfit package.

FYI - As per the doc, [SensorKit places a 24-hour holding period on newly recorded data before an app can access it.](https://developer.apple.com/documentation/sensorkit/srfetchrequest?changes=_5__11&language=objc) So our app won't get any data of this holding period (last 24 hours).

**Comment by avaidyam** [comment]
Thanks, that's helpful to know!

**Comment by jijopulikkottil** [comment]
>  4 (iv) Convert the NSObject subclass into Codable type or directly serialize to JSON data.
> 
> - Code such as data.path_to_space = sample.pathToSpace or data.pathToSpace = sample.pathToSpace where each sample property is manually converted to JSON is UNACCEPTABLE.
> 
> - This conversion must be automatic and apply to any NSObject subclass without knowing its content (properties/keys and values).

In the above [comment](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/139#issuecomment-944364602, ), we are instructed to do it as automatic without knowing the properties. To accomplish this, we have to implement Objective C runtime and KVC. It might be complex as it have inner objects and we won't have control over the key names.

Currently we make each `Sample` as Encodable by specifying the keys. Hope we can continue this way.

```
extension SRAmbientLightSample: Encodable {
    private enum CodingKeys: String, CodingKey {
        case chromaticity
        case lux
        case placement
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(chromaticity, forKey: .chromaticity)
        try container.encode(lux, forKey: .lux)
        try container.encode(placement.rawValue, forKey: .placement)
    }
}
```

**Comment by avaidyam** [comment]
For the initial version, that is fine. We can refine the implementation later.

**Comment by jijopulikkottil** [comment]
@Linoy339 please add the sensor specs mentioned [here](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/139#issuecomment-944364602) (there are 9 new specs. 'com.apple.sensorkit.ambient_light',..etc)

**Comment by jijopulikkottil** [comment]
@avaidyam As per the [steps](https://developer.apple.com/documentation/sensorkit/configuring_your_project_for_sensor_reading), we need to specify the usage description and privacy policy URL. Please provide the same.

1. NSSensorKitUsageDescription
2. NSSensorKitPrivacyPolicyURL
3. Description for each item (ambient_light, ambient_light, etc)

For other privacy item's (say bluetooth) description, we given like "Your bluetooth data is required for clinical study".  Hope we can follow the same for (3).

**Comment by avaidyam** [comment]
@jijopulikkottil Yes - for all of the above items, please say "Your _____ data is required for a clinical research study you have opted to join."

**Comment by jijopulikkottil** [comment]
@avaidyam In plist, we have given 
`NSSensorKitPrivacyPolicyURL = https://docs.lamp.digital/privacy/`

We can push the code once verify the data collection as per latest timestamp.

**Comment by jijopulikkottil** [comment]
@avaidyam among these sensorkit data, only `ambient_light` data has the high frequency (about 10 record per second), hope we can implement frequency settings to this `ambient_light` to manage it.

**Comment by avaidyam** [comment]
@jijopulikkottil Please add support for the `frequency` and `cellular_upload` parameters for all of these sensors.

**Comment by jijopulikkottil** [comment]
@avaidyam , For ambient_light, we set default frequency as 1. hope this is fine rather than system default 10 (about 10 records getting per second when we fetch).

**Comment by jijopulikkottil** [comment]
We have pushed the code. 
The` wrist-state ` data collection is still not passed the QA-testing. We will update soon.

**Comment by avaidyam** [comment]
@jijopulikkottil Please merge the SensorKit code ASAP. I'm closing this issue as we've previously discussed this issue is completed now.

**Comment by jijopulikkottil** [comment]
SensorKit code is already merged and available with latest beta release. [the issue is, we have not received the Wrist-State data yet]

**Comment by jijopulikkottil** [comment]
Currently following sensors are available on staging to Add:

com.apple.sensorkit.ambient_light
com.apple.sensorkit.wrist_state
com.apple.sensorkit.visits
com.apple.sensorkit.device_usage
com.apple.sensorkit.messages_usage
com.apple.sensorkit.phone_usage

**Comment by avaidyam** [comment]
@jijopulikkottil This is great, thank you! @carlan1 is currently testing these on staging. At this time, do these sensors come with any configurable parameters?

**Comment by jijopulikkottil** [comment]
@avaidyam we can set frequency for `com.apple.sensorkit.ambient_light`. [Default is 1 if not specified.](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/139#issuecomment-1149421546)
